iceoryx_hoofs/source/cxx/spinator.cpp

```diff
@@ -0,0 +1,51 @@
+// Copyright (c) 2022 by Apex.AI Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "iceoryx_hoofs/internal/cxx/spinator.hpp"
+
+#include <thread>
+
+namespace iox
+{
+namespace cxx
+{
+namespace internal
+{
+constexpr std::chrono::microseconds spinator::INITIAL_WAITING_TIME;
+constexpr std::chrono::milliseconds spinator::FINAL_WAITING_TIME;
+constexpr uint64_t spinator::YIELD_REPETITIONS;
+constexpr uint64_t spinator::INITIAL_REPETITIONS;
+
+void spinator::yield() noexcept
+{
+    ++m_yieldCount;
```

Comment:
Is this ever reset? (except for overflow, which is another matter).

In a naive implementation yield multiple times until we finally succeed. Next time we yield we could A) start over at 0 yields or B) somehow use the sleep time we reached before. The problem is that B) is only sensible if the time can go down again, which it currently looks like it cannot.

Now as I understand it the sleep time will only go up. 

Phase 1: yield YIELD_REPETITIONS
Phase 2: sleep INITIAL_WAITING_TIME for at most INITIAL_REPETITIONS 
Phase 3: always sleep for FINAL_WAITING_TIME

This is to coarse I think because it will kill latency since if we reach phase 3 we will always sleep FINAL_WAITING_TIME, even if there is not much going on and a yield or shorter sleep would suffice. That is why I would like a reset (which will probably not solve the initial problem), or better, **gradually scale down the sleep time again if we do not sleep often** until success.

The scaling can be something between INITIAL_WAIT_TIME and FINAL_WAITING_TIME and follow an exponential growth/decay. Now we can also add this later and start with this, the main problem is just that initial failures/blocking will increase the sleep time forever which has impacts on latency (we block longer than needed).

Reply:
I would add this later. To come up here with some useful numbers and a strategy we need to define our use case first and come up with a benchmark.

My rough thinking was. We use this currently in the blocking publisher. In the first phase we want to keep the latency low and that is why we yield but after that the latency is already high and the relative amount the yield strategy then adds is less than the latency caused by the subscriber.
If one really needs a reset one can always write:
```
myAdaptiveWait = adaptive_wait();
```

Also added to the doxygen docu.